### PR TITLE
fix: [NPM-WIN] ability to reassign Pod associated with Endpoint

### LINKED ISSUE #1729: Edge Case for memory-starved Windows Server 2022 Nodes with NPM
In memory-starved Windows Server 2022 nodes, there's a rare issue where a Pod may have unexpected connectivity.

### Scenario
- Pod A exists
- At the same time:
  - Pod A is deleted/restarts
  - NPM restarts
  - Pod B comes up with the same IP

Controller events can be jumbled. The possible sequences are:
- Pod A create --> Pod A cleanup --> Pod B create
- Pod A create --> Pod B create --> Pod A cleanup
- Pod B create --> Pod A create --> Pod A cleanup

The first sequence is typical, but the second and third need extra logic.

### Mitigation
If a Pod encounters this issue, restore proper connectivity by bumping the Pod (deleting/recreating).

### Code Fix
This happens because the newly created NPM doesn't know about Pod A's original HNS Endpoint. So NPM can incorrectly associate an Endpoint with a deleted Pod/IP pair.

#### Fix
Use this trait about current code: throws an error if an Endpoint was previously assigned to a Pod, but another Pod claims to have the same IP/Endpoint. The `UpdatePod` will remain in the dirty cache and be retried later.

If a Pod is marked for delete (i.e. when the Pod Metadata has `nodeName=""`), do the following:
- if the Pod is associated with the IP's Endpoint:
    - reset the Endpoint's NPM ACLs
    - mark the Endpoint's Pod as unspecified
- always (this code already exists):
    - remove that Pod Key's `UpdatePod` from the dirty cache so we don't retry with it anymore

##### Sequences 
- *Pod A create --> Pod A cleanup --> Pod B create*
    - Pod A create: Endpoint's Pod is unspecified. Pod A incorrectly assigned to endpoint. Any ACLs for Pod A are added.
    - Pod A cleanup: Reset ACLs for the Endpoint. Mark the Endpoint's Pod as unspecified.
    - Pod B create: Endpoint's Pod is unspecified. Pod B correctly assigned to endpoint. Any ACLs for Pod B are added.
- *Pod A create --> Pod B create --> Pod A cleanup*
    - Pod A create: Endpoint's Pod is unspecified. Pod A incorrectly assigned to endpoint. Any ACLs for Pod A are added.
    - Pod B create: failure since B has the same IP/Endpoint as Pod A. Keep B's `UpdatePod` in the cache, which will be retried next `ApplyDataplane()`
    - Pod A cleanup: Reset ACLs for the Endpoint. Mark the Endpoint's Pod as unspecified.
    - Pod B retry: Endpoint's Pod is unspecified. Pod B correctly assigned to endpoint. Any ACLs for Pod B are added.
- *Pod B create --> Pod A create --> Pod A cleanup*
    - Pod B create: Endpoint's Pod is unspecified. Pod B correctly assigned to endpoint. Any ACLs for Pod B are added.
    - Pod A create: failure since A has the same IP/Endpoint as Pod B. Keep A's `UpdatePod` in the cache, which will be retried next `ApplyDataplane()`
    - Pod A cleanup: No-Op. Pod A `UpdatePod` will be removed from cache so that it's not retried anymore.

### PR DESCRIPTION
**Pull request by huntergregory**
Fix #1729. Improves on Pod markedForDelete idea in original commits of #1503 (see https://github.com/Azure/azure-container-networking/pull/1503/commits/398c99e4d1d2fbbf013a4429a68bb065e79215be).

# Overview
NPM may incorrectly assign a Pod to an Endpoint. NPM needs to be able to reconcile this.

## Scenario
Pod x/a, x/b, and NPM all restart at same time. Pod x/a had IP 10.224.0.1 before restart. Pod x/b has 10.224.0.1 after restart.

NPM can receive the following sequence of events:
1. Pod x/a create --> Pod x/a delete --> Pod x/b create
2. Pod x/a create --> Pod x/b create --> Pod x/a delete
3. ~Pod x/b create --> Pod x/a create --> Pod x/a delete~ (considered impossible)

## Changes
When receiving Pod B event, DP must make sure there are no Pod A ACLs on the Endpoint. Also, DP must not process Pod A events for the same IP afterwards.

### Minor Changes
Remove complex, unnecessary stalePodKey logic. Stale Pod Key was designed to handle this (impossible) situation: a Pod's Update event comes in after its Delete event. It is safe to assume that the Pod Controller won't send an Update after a Delete event for the same Pod/IP pair.

### CODE REVIEW

File: npm/pkg/dataplane/dataplane.go [hunk 1]
```diff
@@ -24,13 +24,30 @@ type Config struct {
 	*policies.PolicyManagerCfg
 }
 
-type updatePodCache struct {
+// dirtyIPCache tracks dirty IPs and Pod updates for those IPs.
+// It tracks IPSets a Running Pod was recently added to or removed from, and which deleted Pods were just deleted from an IP.
+// assumption: a Pod won't take up its previously used IP when restarting (see https://stackoverflow.com/questions/52362514/when-will-the-kubernetes-pod-ip-change)
+// A Pod may be associated with more than one dirtyIP after Pod restarts. The number  some ApplyDataPlane fails several times or ApplyDataPlane is asynchronous.
```
Comment (huntergregory) [inline npm/pkg/dataplane/dataplane.go:30]: TODO: revise end of this comment

File: npm/pkg/dataplane/dataplane_windows.go [hunk 2]
```diff
@@ -113,41 +113,78 @@ func (dp *DataPlane) shouldUpdatePod() bool {
...
+		return nil
+	}
+
 	// Check if pod is already present in cache
 	endpoint, ok := dp.endpointCache.cache[pod.PodIP]
 	if !ok {
 		// ignore this err and pod endpoint will be deleted in ApplyDP
 		// if the endpoint is not found, it means the pod is not part of this node or pod got deleted.
-		klog.Warningf("[DataPlane] did not find endpoint with IPaddress %s for pod %s", pod.PodIP, pod.PodKey)
+		klog.Warningf("[DataPlane] ignoring pod update since there is no corresponding endpoint. IP: %s. podKey: %s", pod.PodIP, pod.PodKey)
```
Comment (huntergregory) [inline npm/pkg/dataplane/dataplane_windows.go:181]: TODO: info, not warning

**Review by vakalapa — APPROVED** [review]

**Pull request merged**